README.md
# neato-mqtt

This program synchronizes your Neato Botvac vacuum robots with an MQTT broker.

It uses the Nucleo API to fetch available robots, read states and send commands. 
The Nucleo API is documented on https://developers.neatorobotics.com/api/nucleo.

## Setup


### Quick start

- Take note of the username and password you use for your Neato account.
- Copy `Settings.example.toml` to `Settings.toml`.
- Edit `Settings.toml` with values matching your setup.
- Try running neato-mqtt with `cargo run`. If your bridge runs recent enough firmware, the program should now launch without errors.

Now you should be able to view your Neato Botvac robots on the MQTT broker, via e.g. [MQTT Explorer](http://mqtt-explorer.com/).

To make the robot(s) clean and other actions, simply publish a message under `home/devices/neato/{id}/set`:

``` json
{
  "action": "StartCleaning", // also works with "startCleaning or "start_cleaning
}
```

If you publish your action under `home/devices/neato/set`, the action will be sent to all robots under `home/devices/neato/` (or based on what `topic` and `set_topic` settings you have in `Settings.toml`).

Available messages are listed on https://developers.neatorobotics.com/api/robot-remote-protocol/housecleaning. 

### Setting Up Mosquitto 

- Ensure Docker is installed and running

- Run mosquitto (with authentication disabled) using:

  ```
  docker run -it -p 1883:1883 eclipse-mosquitto mosquitto -c /mosquitto-no-auth.conf
  ```

### Setting Up MQTT Explorer

- Install [MQTT Explorer](http://mqtt-explorer.com/)

- Connect to the MQTT broker by configuring the connection

  - Setting the protocol to mqtt://
  - Setting the host to localhost
  - Setting the port to 1883

## Topics

The default MQTT topics are as follows:

- `/home/devices/neato/{id}`: Current state of the device serialized as JSON
- `/home/devices/neato/{id}/set`: Sets state of the light to given JSON

## State messages

Example robot state in JSON:

```json
{
  "mac_address": "123456789012",
  "model": "BotVacD6Connected",
  "name": "Vacuum",
  "nucleo_url": "https://nucleo.neatocloud.com:4443",
  "serial": "12345678-123456789012",
  "state": {
    "alert": null,
    "error": null,
    "details": {
      "isCharging": true,
      "isDocked": true,
      "isScheduleEnabled": true,
      "dockHasBeenSeen": false,
      "charge": 99
    },
    "state": 1,
    "action": 0
  }
}
```


CHANGELOG.md
# Changelog

## 0.1.0 (2023-10-30)


### Features

* Action messages published under `/set` for the robots works ([60eaef2](https://github.com/Pingviinituutti/neato-mqtt/commit/60eaef2e5096ec4747b6ea9c46e308761bcb781d))
* adds aliases to mqtt action strings ([95d86ca](https://github.com/Pingviinituutti/neato-mqtt/commit/95d86ca7ed3b2ef9f1fbf94c4d1e03afbb222e9b))
* doing action under device root for `/set` sends command to all robots ([31c41a0](https://github.com/Pingviinituutti/neato-mqtt/commit/31c41a080a6360a3aa2e8465b0f5a7c83d275ea0))
* makes default configs better and adds `MQTT_HOST` env override ([444d00d](https://github.com/Pingviinituutti/neato-mqtt/commit/444d00d612ad2d7f040a3cdb992185faee4ebfe8))


### Bug Fixes

* deletes forgotten test lines ([20a1fce](https://github.com/Pingviinituutti/neato-mqtt/commit/20a1fce1ab53a5abc3ff547c85f642c3801528d4))
* Update release-please.yaml ([e539e1b](https://github.com/Pingviinituutti/neato-mqtt/commit/e539e1b27762dae3fc24d222bab40b3950e52ffd))

## Changelog


src/mqtt.rs
use eyre::Result;
use rand::{distributions::Alphanumeric, Rng};
use rumqttc::{AsyncClient, ConnectionError, Event, MqttOptions, Publish, QoS};
use serde::{Deserialize, Serialize};
use std::{sync::Arc, time::Duration};
use tokio::{sync::watch::Receiver, task};

use log::{debug, error};

use crate::{neato::RobotCmd, settings::MqttSettings};

#[derive(Clone, Deserialize, Serialize, Debug)]
pub struct MqttSetMessage {
    pub action: RobotCmd,
}

#[derive(Clone, Deserialize, Serialize, Debug)]
pub struct SendAction {
    pub id: String,
    pub action: RobotCmd,
}

#[derive(Clone)]
pub struct MqttClient {
    pub client: AsyncClient,
    pub rx: Receiver<Option<SendAction>>,
    pub topic: String,
    pub set_topic: String,
    pub settings: MqttSettings,
}

pub fn get_id_from_topic(topic: &String, set_topic: &str) -> Result<String> {
    if let Some((start, end)) = set_topic.split_once("{id}") {
        Ok(topic.replace(start, "").replace(end, ""))
    } else {
        error!("Could not get id from topic: '{}'", topic);
        Err(eyre::eyre!("Could not get id from topic: '{}'", topic))
    }
}

struct NotificationResult {
    message: Option<Publish>,
}

async fn handle_notification(
    client: &AsyncClient,
    notification: Result<Event, ConnectionError>,
    mqtt_settings: &MqttSettings,
) -> Result<NotificationResult> {
    debug!("Notification: {:?}", notification);
    match notification? {
        rumqttc::Event::Incoming(rumqttc::Packet::ConnAck(_)) => {
            client
                .subscribe(mqtt_settings.get_set_topic_with_wildcard(), QoS::AtMostOnce)
                .await?;
            client
                .subscribe(mqtt_settings.get_topic_with_id_as_set(), QoS::AtMostOnce)
                .await?;
            // return Ok(NotificationResult{message: None})
        }
        rumqttc::Event::Incoming(rumqttc::Packet::Publish(msg)) => {
            return Ok(NotificationResult { message: Some(msg) })
        }
        _ => {}
    }

    Ok(NotificationResult { message: None })
    // Err(eyre::eyre!("Could not get message"))
}

pub async fn init(mqtt_settings: &MqttSettings) -> Result<MqttClient> {
    let random_string: String = rand::thread_rng()
        .sample_iter(&Alphanumeric)
        .take(8)
        .map(char::from)
        .collect();

    let mut options = MqttOptions::new(
        format!("{}-{}", mqtt_settings.id.clone(), random_string),
        mqtt_settings.host.clone(),
        mqtt_settings.port,
    );
    options.set_keep_alive(Duration::from_secs(5));
    let (client, mut eventloop) = AsyncClient::new(options, 10);

    let subscribe_settings = mqtt_settings.clone();
    let subscribe_client = client.clone();

    let config_clone = Arc::new(mqtt_settings.clone());

    let (tx, rx) = tokio::sync::watch::channel(None);

    // Listen on set_topic, for example `home/devices/neato/{id}/set`
    task::spawn(async move {
        loop {
            let notification = eventloop.poll().await;

            let id = subscribe_settings.id.clone();
            let config_clone = Arc::clone(&config_clone);

            let res =
                handle_notification(&subscribe_client, notification, &subscribe_settings.clone())
                    .await;

            match res {
                Ok(NotificationResult { message: Some(msg) }) => {
                    debug!("Reveiced MQTT Publish for topic: {:?}", &msg.topic);
                    let id = get_id_from_topic(&msg.topic, &config_clone.set_topic).unwrap();
                    debug!("Id is: {:?}", id);
                    let payload: MqttSetMessage = match serde_json::from_slice(&msg.payload) {
                        Ok(pl) => pl,
                        Err(e) => {
                            error!("Could not parse JSON payload: {:?}", e);
                            continue;
                        }
                    };
                    debug!("Payload is: {:?}", payload);
                    let device = {
                        SendAction {
                            id,
                            action: payload.action,
                        }
                    };
                    tx.send(Some(device)).expect("Failed to send message");
                }
                Err(e) => {
                    error!(
                        target: &id.to_string(),
                        "MQTT error: {:?}", e
                    );
                    tokio::time::sleep(Duration::from_secs(1)).await;
                }
                _ => {}
            }
        }
    });

    // listen on the topic set, for example `home/devices/neato/set`
    // If an action is received here, the command should affect all devices
    // TODO: implement this

    Ok(MqttClient {
        client,
        rx,
        topic: mqtt_settings.topic.clone(),
        set_topic: mqtt_settings.set_topic.clone(),
        settings: mqtt_settings.clone(),
    })
}


src/neato.rs
use std::{
    fmt,
    sync::{Arc, Mutex},
    time::Duration,
};

use tokio::sync::Mutex as AsyncMutex;

use chrono::Utc;
use color_eyre::Result;
use eyre::eyre;
use hmac::{Hmac, Mac};
use serde::{Deserialize, Serialize};
use sha2::Sha256;

use log::{debug, error, info};

use crate::{mqtt::MqttClient, settings::NeatoSettings};
use crate::{
    mqtt::SendAction,
    neato_types::{HouseCleaningParams, NeatoState, PublicRobot, Robot, RobotMessage},
};

impl Robot {
    pub async fn publish(&self, mqtt_client: MqttClient) -> color_eyre::Result<()> {
        let topic = mqtt_client
            .topic
            .clone()
            .replace("{id}", self.name.as_str());

        let v = serde_json::to_value(self).unwrap();
        let public_robot: PublicRobot = serde_json::from_value(v).unwrap();
        mqtt_client
            .client
            .publish(
                topic,
                rumqttc::QoS::AtMostOnce,
                false,
                serde_json::to_string(&public_robot)?,
            )
            .await
            .unwrap();

        Ok(())
    }
}

#[derive(Deserialize)]
struct SessionsResponse {
    access_token: String,
}

#[derive(Serialize)]
struct AuthBody {
    email: String,
    password: String,
}

const BASE_URL: &str = "https://beehive.neatocloud.com";

type HmacSha256 = Hmac<Sha256>;

#[derive(Clone, PartialEq, Deserialize, Serialize, Debug)]
pub enum RobotCmd {
    #[serde(alias = "startCleaning", alias = "start_cleaning")]
    StartCleaning,
    #[serde(alias = "stopCleaning", alias = "stop_cleaning")]
    StopCleaning,
    #[serde(alias = "pauseCleaning", alias = "pause_cleaning")]
    PauseCleaning,
    #[serde(alias = "resumeCleaning", alias = "resume_cleaning")]
    ResumeCleaning,
    #[serde(alias = "sendToBase", alias = "send_to_base")]
    SendToBase,
    GetRobotState,
}

impl fmt::Display for RobotCmd {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            RobotCmd::StartCleaning => write!(f, "startCleaning"),
            RobotCmd::StopCleaning => write!(f, "stopCleaning"),
            RobotCmd::PauseCleaning => write!(f, "pauseCleaning"),
            RobotCmd::ResumeCleaning => write!(f, "resumeCleaning"),
            RobotCmd::SendToBase => write!(f, "sendToBase"),
            RobotCmd::GetRobotState => write!(f, "getRobotState"),
        }
    }
}

impl RobotCmd {
    pub fn build_robot_message(&self) -> RobotMessage {
        match self {
            RobotCmd::StartCleaning => RobotMessage {
                req_id: String::from("77"),
                cmd: String::from("startCleaning"),
                params: Some(HouseCleaningParams {
                    category: 4,
                    mode: 1,
                    navigation_mode: 2,
                }),
            },
            other => RobotMessage {
                req_id: String::from("77"),
                cmd: other.to_string(),
                params: None,
            },
        }
    }
}

type SharedRobots = Arc<AsyncMutex<Vec<Robot>>>;

#[derive(Clone)]
pub struct Neato {
    mqtt_client: MqttClient,
    settings: NeatoSettings,
    robots: SharedRobots,
    last_state_update: Arc<Mutex<Option<chrono::DateTime<chrono::Utc>>>>,
}

impl Neato {
    pub fn new(mqtt_client: MqttClient, neato_settings: &NeatoSettings) -> Neato {
        Neato {
            mqtt_client,
            settings: neato_settings.clone(),
            robots: Arc::new(AsyncMutex::new(Vec::new())),
            last_state_update: Arc::new(Mutex::new(None)),
        }
    }

    // pub async fn get_robot_names(&self) -> Vec<String> {
    //     self.robots.lock().await.iter().map(|r| r.name.clone()).collect()
    // }

    pub async fn init(mut self) -> color_eyre::Result<Neato> {
        info!("Initializing Neato cloud integration");
        self.robots = Arc::new(AsyncMutex::new(get_robots(&self.settings.clone()).await?));
        // let robots_with_states = update_robot_states(get_robots(neato_settings).await?).await?;

        for robot in self.robots.lock().await.iter() {
            let _r = robot.clone();
            info!("Found robot: {:?}", robot.name);
            debug!("Robot info: {:?}", robot);
        }

        // Start the state polling loop
        match self.init_polling().await {
            Ok(_) => (),
            Err(err) => {
                error!("Error initializing polling: {}", err);
            }
        }

        match self.init_react_to_subscription_messages().await {
            Ok(_) => (),
            Err(err) => {
                error!("Error initializing subscription messages: {}", err);
            }
        };

        info!("Neato connection initialized");

        Ok(self)
    }

    async fn force_update_states(&self) -> color_eyre::Result<()> {
        for robot in self.robots.lock().await.iter_mut() {
            debug!("Robot info before update: {:?}", robot);

            let result = send_command(robot, &RobotCmd::GetRobotState).await?;
            let serialized_result: NeatoState = serde_json::from_str(&result).unwrap();
            robot.state = Some(serialized_result);
            // robot.state = robot_map.get(&robot.serial).unwrap().state.clone();
            debug!("Robot info after update: {:?}\n", robot);
        }

        *self.last_state_update.lock().unwrap() = Some(Utc::now());

        Ok(())
    }

    async fn update_states(&self) -> color_eyre::Result<()> {
        // don't update states if we've done it recently
        // let last_update = self.last_state_update.lock().unwrap().clone();
        match self.last_state_update.try_lock() {
            Ok(last_state_update) => {
                if last_state_update.is_some()
                    && (Utc::now() - last_state_update.unwrap()).num_seconds()
                        < self.settings.cache_timeout as i64
                {
                    let ms_ago = (Utc::now() - last_state_update.unwrap()).num_milliseconds();
                    info!(
                        "Skipping update, last update was only {:.2} seconds ago.",
                        ms_ago as f64 / 1000.0
                    );
                    return Ok(());
                } else {
                    debug!("Updating robot states");
                }
            }
            Err(_) => {
                info!("Skipping update, something else seems to be updating the state");
                return Ok(());
            }
        }

        // Now we lock the robots again and update the state of each robot
        self.force_update_states().await?;

        Ok(())
    }

    async fn init_polling(&self) -> color_eyre::Result<()> {
        let poll_rate = Duration::from_millis(self.settings.poll_interval as u64 * 1000);
        let neato = self.clone();
        let mqtt_client = self.mqtt_client.clone();

        tokio::spawn(async move {
            let mut interval = tokio::time::interval(poll_rate);

            loop {
                interval.tick().await;
                match neato.update_states().await {
                    Ok(_) => (),
                    Err(err) => {
                        info!("Error updating robot states: {}", err);
                    }
                };

                for robot in neato.robots.lock().await.iter() {
                    robot.publish(mqtt_client.clone()).await.unwrap();
                }
            }
        });
        Ok(())
    }

    async fn init_react_to_subscription_messages(&self) -> color_eyre::Result<()> {
        let mut s = self.clone();

        tokio::spawn(async move {
            loop {
                s.mqtt_client
                    .rx
                    .changed()
                    .await
                    .expect("Expected rx channel never to close");
                let msg = s.mqtt_client.rx.borrow().clone();

                println!("Received update instruction! Device: {:?}", msg);

                match msg {
                    Some(SendAction {
                        action: RobotCmd::GetRobotState,
                        ..
                    }) => {
                        debug!("We don't do state updates from set messages");
                    }
                    Some(SendAction { action, id }) => {
                        let robots = match id.as_str() {
                            // if id is "set", send to all robots
                            "set" => s.robots.lock().await.clone(),
                            _ => {
                                // Filter out all robots that don't match the id
                                s.robots
                                    .lock()
                                    .await
                                    .clone()
                                    .into_iter()
                                    .filter(|r| r.name == id)
                                    .collect()
                            }
                        };

                        info!("Sending command: {}", action);
                        if robots.is_empty() {
                            error!(
                                "No robots found with name \"{}\". Aborting sending command",
                                id
                            );
                            continue;
                        }
                        info!(
                            "Affected robots: {:?}",
                            robots
                                .iter()
                                .map(|r| r.name.clone())
                                .collect::<Vec<String>>()
                        );
                        if s.settings.dry_run {
                            info!("Setting neato.dry_run enabled, not sending command");
                        } else {
                            for robot in robots {
                                send_command(&robot, &action).await.unwrap();
                            }
                        }
                    }
                    _ => {}
                };
            }
        });

        Ok(())
    }
}

async fn get_robots(config: &NeatoSettings) -> Result<Vec<Robot>> {
    let body = AuthBody {
        email: config.email.clone(),
        password: config.password.clone(),
    };

    let token = surf::post(&format!("{}/sessions", BASE_URL))
        .body(surf::Body::from_json(&body).map_err(|err| eyre!(err))?)
        .await
        .map_err(|err| eyre!(err))?
        .body_json::<SessionsResponse>()
        .await
        .map_err(|err| eyre!(err))?
        .access_token;

    let robots = surf::get(&format!("{}/users/me/robots", BASE_URL))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await
        .map_err(|err| eyre!(err))?
        // .body_string() // in case you want to debug the whole response
        .body_json::<Vec<Robot>>()
        .await
        .map_err(|err| eyre!(err))?;

    Ok(robots)
}

async fn send_command(robot: &Robot, cmd: &RobotCmd) -> Result<String> {
    // https://developers.neatorobotics.com/api/nucleo
    let robot_message = cmd.build_robot_message();

    debug!(
        "Robot name {}, Sending command: {:?}",
        robot.name, robot_message
    );

    let body = serde_json::to_string(&robot_message)?;
    let serial = robot.serial.to_lowercase();
    let date: String = format!("{}", Utc::now().format("%a, %d %b %Y %H:%M:%S GMT"));
    let string_to_sign = format!("{}\n{}\n{}", serial, date, body);

    // Create HMAC-SHA256 instance which implements `Mac` trait
    let mut mac = HmacSha256::new_from_slice(robot.secret_key.as_bytes())
        .expect("HMAC can take key of any size");
    mac.update(string_to_sign.as_bytes());

    let signature = hex::encode(mac.finalize().into_bytes());

    let result = surf::post(&format!(
        "{}/vendors/neato/robots/{}/messages",
        robot.nucleo_url, robot.serial
    ))
    .header("Accept", "application/vnd.neato.nucleo.v1")
    .header("Date", date)
    .header("Authorization", format!("NEATOAPP {}", signature))
    .body(surf::Body::from_json(&robot_message).map_err(|err| eyre!(err))?)
    .await
    .map_err(|err| eyre!(err))?
    .body_string()
    .await
    .map_err(|err| eyre!(err))?;

    debug!("response: {}", result);

    Ok(result)
}


src/neato_types.rs
use std::fmt;

use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

#[derive(Clone, Deserialize, Serialize, Debug)]
pub struct Robot {
    pub mac_address: String,
    pub model: String,
    pub name: String,
    pub nucleo_url: String,
    pub secret_key: String,
    pub serial: String,
    pub state: Option<NeatoState>,
}

#[derive(Clone, Deserialize, Serialize, Debug)]
pub struct PublicRobot {
    pub mac_address: String,
    pub model: String,
    pub name: String,
    pub nucleo_url: String,
    pub serial: String,
    pub state: Option<NeatoState>,
}

#[derive(Serialize, Debug)]
pub struct HouseCleaningParams {
    /// Should be set to 4 for persistent map
    /// 2 without persistent map
    pub category: u32,

    /// 1 is eco, 2 is turbo
    pub mode: u32,

    /// 1 is normal, 2 is extra care, 3 is deep. 3 requires mode = 2.
    #[serde(rename = "navigationMode")]
    pub navigation_mode: u32,
}

#[derive(Serialize, Debug)]
pub struct RobotMessage {
    #[serde(rename = "reqId")]
    pub req_id: String,
    pub cmd: String,
    pub params: Option<HouseCleaningParams>,
}

#[derive(Clone, Deserialize, Serialize, Debug)]
pub struct RobotStateDetails {
    #[serde(rename = "isCharging")]
    pub is_charging: bool,
    #[serde(rename = "isDocked")]
    pub is_docked: bool,
    #[serde(rename = "isScheduleEnabled")]
    pub is_schedule_enabled: bool,
    #[serde(rename = "dockHasBeenSeen")]
    pub dock_has_been_seen: bool,
    pub charge: i8,
}

#[derive(Clone, Copy, Serialize_repr, Deserialize_repr, Debug, PartialEq)]
#[repr(u8)]
pub enum RobotState {
    Invalid = 0,
    Idle = 1,
    Busy = 2,
    Paused = 3,
    Error = 4,
}

impl fmt::Display for RobotState {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
        // or, alternatively:
        // fmt::Debug::fmt(self, f)
    }
}

// https://developers.neatorobotics.com/api/robot-remote-protocol/request-response-formats#13-6-strong-code-action-code-em-integer-em-strong
// action: integer
// If the state is busy, this element specifies what the robot is or has been busy doing.
// If the state is pause or error, it specifies the activity that the Robot was doing.
// If state is other, this element is null.
#[derive(Clone, Copy, Serialize_repr, Deserialize_repr, Debug)]
#[repr(u8)]
pub enum RobotAction {
    Invalid = 0,
    HouseCleaning = 1,
    SpotCleaning = 2,
    ManualCleaning = 3,
    Docking = 4,
    UserMenuActive = 5,
    SuspendedCleaning = 6,
    Updating = 7,
    CopyingLogs = 8,
    RecoveringLocation = 9,
    IECTest = 10,
    MapCleaning = 11,
    ExploringMap = 12,
    AcquiringPersisntentMapIDs = 13,
    CreatingAndUploadingMap = 14,
    SuspendedExploration = 15,
}

impl fmt::Display for RobotAction {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
        // or, alternatively:
        // fmt::Debug::fmt(self, f)
    }
}

#[derive(Clone, Deserialize, Serialize, Debug)]
pub struct NeatoState {
    pub alert: Option<String>,
    pub error: Option<String>,
    pub details: RobotStateDetails,
    pub state: RobotState,
    pub action: RobotAction,
}


src/settings.rs
use std::env;

use log::log_enabled;
use serde::Deserialize;

#[derive(Clone, Deserialize, Debug)]
pub struct NeatoSettings {
    pub email: String,
    pub password: String,
    pub poll_interval: u16, // seconds
    pub cache_timeout: u16, // seconds
    pub decode_state: bool,
    pub dry_run: bool,
}

fn default_poll_interval() -> u16 {
    if log_enabled!(log::Level::Debug) {
        5 // seconds
    } else {
        60 // 1 minute
    }
}

fn default_cache_timeout() -> u16 {
    if log_enabled!(log::Level::Debug) {
        30 // seconds
    } else {
        5 * 60 // 5 minutes
    }
}

#[derive(Clone, Deserialize, Debug)]
pub struct MqttSettings {
    pub id: String,
    pub host: String,
    pub port: u16,
    pub topic: String,
    pub set_topic: String,
}

impl MqttSettings {
    // pub fn get_topic_with_wildcard(&self) -> String {
    //     // Return the topic with `{id}` replaced with `+`
    //     self.topic.replace("{id}", "+")
    // }
    pub fn get_topic_with_id_as_set(&self) -> String {
        self.topic.replace("{id}", "set")
    }
    // pub fn get_topic_without_id(&self) -> String {
    //     // Return the topic with `{id}` removed
    //     self.topic.replace("{id}", "")
    // }
    pub fn get_set_topic_with_wildcard(&self) -> String {
        // Return the topic with `{id}` replaced with `+`
        self.set_topic.replace("{id}", "+")
    }
}

#[derive(Clone, Deserialize, Debug)]
pub struct Settings {
    pub neato: NeatoSettings,
    pub mqtt: MqttSettings,
}

pub fn read_settings() -> Result<Settings, config::ConfigError> {
    config::Config::builder()
        .add_source(config::File::with_name("Settings"))
        .set_default("mqtt.host", "localhost")?
        .set_default("mqtt.port", 1883)?
        .set_default("mqtt.topic", "home/devices/neato/{id}")?
        .set_default("mqtt.set_topic", "home/devices/neato/{id}/set")?
        .set_default("neato.poll_interval", default_poll_interval())?
        .set_default("neato.cache_timeout", default_cache_timeout())?
        .set_default("neato.decode_state", false)?
        .set_default("neato.dry_run", false)?
        .set_override_option("mqtt.host", env::var("MQTT_HOST").ok())?
        .build()?
        .try_deserialize::<Settings>()
}
